src/Snak/SnakList.php

```diff
@@ -15,17 +19,270 @@
  * @author Jeroen De Dauw < jeroendedauw@gmail.com >
  * @author Addshore
  */
-class SnakList extends HashArray {
+class SnakList extends ArrayObject implements Hashable, Comparable {
 
 	/**
-	 * @see GenericArrayObject::getObjectType
+	 * Maps snak hashes to their offsets.
+	 *
+	 * @var array [ snak hash (string) => array [ snak offset (string|int) ] | snak offset (string|int) ]
+	 */
+	private $offsetHashes = [];
+
+	/**
+	 * @var integer
+	 */
+	private $indexOffset = 0;
+
+	/**
+	 * @see ArrayObject::__construct
+	 *
+	 * @param array|Traversable|null $input
+	 * @param int $flags
+	 * @param string $iteratorClass
+	 *
+	 * @throws InvalidArgumentException
+	 */
+	public function __construct( $input = null, $flags = 0, $iteratorClass = 'ArrayIterator' ) {
```

Comment:
In my opinion this constructor is exposing to much (undocumented) complexity that is not needed by users of the `SnakList` class. Please tell me if I miss something. Otherwise I suggest to merge #439 first and rebase this patch.